### ISSUE TITLE
KV WatchAll takes unexpectedly long in a specific situation

### ISSUE THREAD

**Issue by pcsegal** [nats-io/nats-server#4987]
### Observed behavior

When using NATS KV `WatchAll`, the time for `WatchAll` to run through all the keys seems to become very long depending on the rate at which the objects were put in the bucket (more details below).

### Expected behavior

It seems that the `WatchAll` time should depend only on the number of entries in the bucket.

### Server and client version

NATS server: v2.10.9
Go client (github.com/nats-io/nats.go): v1.32.0

### Host environment

OS: Ubuntu 22.04.1 LTS on WSL2 (Windows 10)
CPU: Intel(R) Core(TM) i3-3220 CPU @ 3.30GHz
RAM: 8.00 GB

### Steps to reproduce

The below script can be used to reproduce this issue (see below for steps).

```go
package main

import (
    "encoding/json"
    "errors"
    "flag"
    "fmt"
    "sync"
    "time"

    "github.com/nats-io/nats.go"
)

type Item struct {
    Data string
}

func getBucket(js nats.JetStreamContext, bucketName string, createIfNotExists bool) (nats.KeyValue, error) { //nolint
    bucket, err := js.KeyValue(bucketName)
    if errors.Is(err, nats.ErrBucketNotFound) && createIfNotExists {
        bucket, err = js.CreateKeyValue(&nats.KeyValueConfig{
            Bucket: bucketName,
        })
    }
    if err != nil {
        return nil, fmt.Errorf("create bucket: %w", err)
    }
    return bucket, nil
}

func watch(kv nats.KeyValue) (int, error) {
    w, err := kv.WatchAll(nats.IgnoreDeletes())
    if err != nil {
        return 0, fmt.Errorf("watch start: %w", err)
    }
    count := 0
    updates := w.Updates()
    for e := range updates {
        if e == nil {
            break
        }
        item := new(Item)
        if err := json.Unmarshal(e.Value(), item); err != nil {
            return 0, fmt.Errorf("unmarshal: %w", err)
        }
        count++
    }
    if err := w.Stop(); err != nil {
        return 0, fmt.Errorf("watch stop: %w", err)
    }
    return count, nil
}

func fillBucket(kv nats.KeyValue, numItems int, sleepBetweenPuts bool) error {
    var wg sync.WaitGroup
    wg.Add(numItems)

    for i := 0; i < numItems; i++ {
        i := i
        go func(wg *sync.WaitGroup) {
            defer wg.Done()
            key := fmt.Sprintf("item-%d0000000000000000", i)
            item := Item{Data: key}
            bytes, err := json.Marshal(item)
            if err != nil {
                fmt.Printf("json marshal: %v\n", err)
                return
            }
            _, err = kv.Put(key, bytes)
            if err != nil {
                fmt.Printf("kv put: %v\n", err)
                return
            }
            _, err = kv.Put(key, bytes)
            if err != nil {
                fmt.Printf("kv put: %v\n", err)
                return
            }
        }(&wg)

		if sleepBetweenPuts {
        	time.Sleep(100 * time.Microsecond)
		}
    }
    wg.Wait()
    return nil
}

func main() {
    n := flag.Int("n", 10000, "number of objects in bucket")
    sleepBetweenPuts := flag.Bool("s", false, "whether to sleep between puts")
    flag.Parse()

    var nc *nats.Conn
    var js nats.JetStreamContext
    nc, err := nats.Connect("127.0.0.1")
    if err != nil {
        panic(err)
    }
    defer nc.Close()
    js, err = nc.JetStream()
    if err != nil {
        panic(err)
    }

    bucketName := "testkvbucket"

    err = js.DeleteKeyValue(bucketName)
    if err != nil && !errors.Is(err, nats.ErrStreamNotFound) {
        panic(err)
    }

    kv, err := getBucket(js, bucketName, true)
    if err != nil {
        panic(err)
    }

    if err := fillBucket(kv, *n, *sleepBetweenPuts); err != nil {
        panic(err)
    }

    time.Sleep(1000 * time.Millisecond)

    var totalTime time.Duration

    for i := 0; i < 5; i++ {
        start := time.Now()
        count, err := watch(kv)
        if err != nil {
            panic(err)
        }
        t := time.Since(start)
        fmt.Printf("watched %d items in %s\n", count, t)
        totalTime += t
    }

    fmt.Printf("average test time: %s\n", totalTime/5)
}
```

This script has two parts:

* It populates the bucket (`fillBucket`) with the given number of objects (the number of items is determined by the `-n` flag). The population of the bucket is done by launching, for each item, a goroutine that makes two NATS KV `Put`s to insert that item into the bucket (NOTE: this issue only manifests itself if we use at least two `Put`s; if we change the script to make only one `Put`, the issue does not happen).

  If the `-s` flag is turned on, then, between launching two goroutines, `fillBucket` sleeps for 100 microseconds. The point of this is to increase the total time over which the data population occurs.
* It watches the bucket, using `WatchAll`. In the script, this is done 5 times. For each run, it measures the time `WatchAll` takes to make the first run through all the objects in the bucket.

In order to reproduce the issue, run the below steps:

1. Run this script for 100000 records, without the `-s` flag. That is: `go run . -n 100000`. The output we obtained was this:

    ```
    watched 100000 items in 465.778172ms
    watched 100000 items in 553.001866ms
    watched 100000 items in 540.506767ms
    watched 100000 items in 567.132065ms
    watched 100000 items in 509.049069ms
    average test time: 527.093587ms
    ```
2. Run this script for 100000 records, with the `-s` flag: `go run . -n 100000 -s`. There was a significant difference in the run time of `WatchAll`, even though the total number of items is the same:

    ```
    watched 100000 items in 25.030743586s
    watched 100000 items in 25.689579883s
    watched 100000 items in 25.458699234s
    watched 100000 items in 1m19.135490707s
    watched 100000 items in 1m40.022924511s
    average test time: 51.067487584s
    ```

NOTES:
1. You may have to repeat step 2 in order see a similar difference.
2. You may have to increase the size of the keys in order to see a difference. In the above script, for the i-th object, the key is `fmt.Sprintf("item-%d0000000000000000", i)`. You may have to increase the number of trailing `0`s to see the effect. For example, we tested this same script in another Ubuntu (not on WSL2), and we had to make keys with 35 zeros (i.e., `fmt.Sprintf("item-%d00000000000000000000000000000000000", i)`) in order to see this run time difference show up.

**Comment by Jarema** [comment]
Hey.  Thanks for detailed explanation. I'm looking into this one.

**Comment by Jarema** [comment]
I was able to reproduce the issue.
It was not related to client behaviour, but rather how adding sleep changed the structure of data in the Stream compared to `Put`s fired in goroutines, which affected speed of processing when consuming the messages. I didn't investigate that one in detail.

However, we did a big improvements in `main` to how we handle subjects.
I retested on the `main` and now it seems that both examples result in very similar times.

Feel free to test on `main` to confirm.
